### PR TITLE
Replace Datasets DatasetSelect component on groups page with ListResources [#870]

### PR DESCRIPTION
**Pull request by genehack**
## Description of proposed changes

Replaces the current "datasets" DatasetSelect component with a ListResources component that hits the same backend `charon` API. Also refactors in `useDataFetch` to change the hard-coded parser into a passed-in (mandatory) callback function, and updates two existing uses to the new pattern. Finally, adds in conditional display logic to the ListResources component to hide or otherwise not set UI elements that don't make sense with the `groups` dataset (e.g., last modified dates, since we don't have those). 

[Preview](https://nextstrain-s-group-reso-fafosr.herokuapp.com/groups)

## Related issue(s)

#870 

## Checklist

- [x] Checks pass
- [x] Check if changes affect the [resource index JSON revision](https://docs.nextstrain.org/projects/nextstrain-dot-org/en/latest/resource-collection.html#resource-index-revisions)

### CODE REVIEW

File: static-site/src/components/ListResources/index.tsx
Comment (victorlin) [inline static-site/src/components/ListResources/index.tsx:126]: The object type `{pathPrefix: string, pathVersions: PathVersions}` is used in many places, I think enough to warrant some standardization.

Suggestion: Add an interface in `types.ts` and use this in place of the object type.

```ts
interface ResourceListingInfo {
  pathPrefix: string
  pathVersions: PathVersions
}
```

File: static-site/src/components/ListResources/index.tsx
Comment (victorlin) [inline static-site/src/components/ListResources/index.tsx:48]: ### Nitpick: naming the callback function

Currently the caller names it `parserCallBackFxn` while the function signature uses `parserCallBack`.

Suggestion 1: use the same name. Between the two, I'd pick `parserCallBack` – the `Fxn` abbreviation is new to this codebase and I think redundant with the type info.

Suggestion 2: Based on interface name suggested in based on https://github.com/nextstrain/nextstrain.org/pull/904#discussion_r1631592549, use the name `getResourceListingInfo`. Usage would look like this, which to me reads better than the current names.

```js
// static-site/src/sections/pathogens.jsx
const getResourceListingInfo = async (response) => (await response.json()).dataset[sourceId];

// static-site/src/components/ListResources/useDataFetch.ts
({ pathVersions, pathPrefix } = await getResourceListingInfo(response));

// static-site/src/pages/groups.jsx
async function getResourceListingInfo(response) {
    const datasets = (await response.json()).datasets;
    // ...
    return { pathPrefix: "groups/", pathVersions };
}
```

File: static-site/src/pages/groups.jsx
Comment (victorlin) [inline static-site/src/pages/groups.jsx:16]: This uses `charon/getAvailable` but it wasn't super obvious to me - there's a big distance between this line and [where the use of charon is defined](https://github.com/nextstrain/nextstrain.org/blob/bc73c3b9336fb48602d918ef41452b69611abd55/static-site/src/components/ListResources/useDataFetch.ts#L35).

Suggestion: define the URL and make the fetch request directly in the definition of each `parserCallBackFxn`, rather than shared across all invocations of `useDataFetch` and conditions written inside of that. This would be a larger refactor, but I think could be in scope of this PR if the goal is to make ListResources re-usable outside of `/pathogens` and `/staging` pages.

File: static-site/src/pages/groups.jsx
Comment (victorlin) [inline static-site/src/pages/groups.jsx:24]: This seems like a hack around fragile logic under `versioned=false` that requires `PathVersions` values to be `string[]` which doesn't mesh well with `/groups`. Ideally it can be `undefined`:

```ts
export interface PathVersions {
  [name: string]: string[] | undefined
}
```

This would be a larger refactor which also necessitates attributes like `Resource.lastUpdated` to be optional, but maybe worth doing now.

File: static-site/src/components/ListResources/IndividualResource.tsx
Comment (victorlin) [inline static-site/src/components/ListResources/IndividualResource.tsx:161]: Suggestion: condition on `resource.lastUpdated` directly

```suggestion
  const description = resource.lastUpdated ? `Last known update on ${resource.lastUpdated}` : "";
```

File: static-site/src/components/ListResources/useDataFetch.ts
Comment (genehack) [reply static-site/src/components/ListResources/useDataFetch.ts:24]: yeah, i've frequently found problems by asserting explicit return types before; tiny bit of hassle, but worth it when it find something funky!

File: static-site/src/components/ListResources/index.tsx
Comment (genehack) [reply static-site/src/components/ListResources/index.tsx:48]: I've got a lean for callbacks to be clearly identifiable by name; additionally, this callback isn't getting, it's only parsing.

How does `parseResourceListingInfoCb` grab you?

File: static-site/src/pages/groups.jsx
Comment (genehack) [reply static-site/src/pages/groups.jsx:16]: I structured this the way I did because it felt like the smallest effective change, and I wasn't sure we really wanted to effectively outsource the production of the ResourceListingInfo datastructure to the consumer of the <ListResources> component — which is where I think this suggestion ends up once the dust settles.

Anybody else have a strong opinion here?

File: static-site/src/pages/groups.jsx
Comment (genehack) [reply static-site/src/pages/groups.jsx:24]: I didn't like the empty string approach, but using `undefined` or `nil` didn't feel great either, and changing all the other code to deal with the value-side _not_ being an array felt like a *lot* of work.

File: static-site/src/components/ListResources/IndividualResource.tsx
Comment (genehack) [reply static-site/src/components/ListResources/IndividualResource.tsx:161]: I want to say I tried this and it didn't DTRT …but I don't remember what it did. Will give it a shot.

File: static-site/src/components/ListResources/index.tsx
Comment (victorlin) [reply static-site/src/components/ListResources/index.tsx:48]: Better – not a huge fan of abbreviations (seems to go against "clearly identifiable by name"), but dealer's choice!

Comment (victorlin) [reply static-site/src/components/ListResources/index.tsx:48]: If spelling it out, I would use `Callback` instead of `CallBack` (with `React.useCallback` in mind)

File: static-site/src/pages/groups.jsx
Comment (victorlin) [reply static-site/src/pages/groups.jsx:24]: > using `undefined` or `nil` didn't feel great either

`undefined` sounds right to me... e.g. "what are the dates associated with this path?"

1. `undefined` – there are no dates!
2. there are no dates, but here's a single date with the value `""`

> changing all the other code to deal with the value-side _not_ being an array felt like a _lot_ of work.

It might be nontrivial, but I would consider it in scope of making ListResources re-usable outside of `/pathogens` and `/staging` pages (same reasoning as https://github.com/nextstrain/nextstrain.org/pull/904#discussion_r1631608768). Otherwise, ListResources will remain tailored for those two "first-class" pages, with pages such as `/groups` considered "second-class" and relying on workarounds to match assumptions made for the first-class pages.

File: static-site/src/components/ListResources/useDataFetch.ts
Comment (jameshadfield) [reply static-site/src/components/ListResources/useDataFetch.ts:24]: P.S. We could use [@typescript-eslint/explicit-function-return-type](https://typescript-eslint.io/rules/explicit-function-return-type/) if we want to enforce this

File: static-site/src/pages/groups.jsx
Comment (jameshadfield) [reply static-site/src/pages/groups.jsx:16]: >  I wasn't sure we really wanted to effectively outsource the production of the ResourceListingInfo datastructure to the consumer of the component — which is where I think this suggestion ends up once the dust settles.

I'm not sure we need to (nor should) go this far. But I do think adding the `fetch` call to the callback (and therefore the API address) is a better separation of concerns.

File: static-site/src/pages/groups.jsx
Comment (jameshadfield) [reply static-site/src/pages/groups.jsx:24]: > Otherwise, ListResources will remain tailored for those two "first-class" pages, with pages such as /groups considered "second-class" and relying on workarounds to match assumptions made for the first-class pages.

This is a compelling point for me. In the short-term future we will probably have:
* 1 page with versions (/pathogens)
* 1 page with last-updated date (/staging)
* lots of pages with no dates (/groups, `/groups/<NAME>`, community splash pages)

And that's not including [a files listing page](https://github.com/nextstrain/nextstrain.org/pull/828) which would itself require larger changes. 

For what it's worth, I'd use an empty array to represent no dates.

File: static-site/src/components/ListResources/index.tsx
Comment (genehack) [reply static-site/src/components/ListResources/index.tsx:47]: Split the difference with `parseResourceListingCallback`.

File: static-site/src/pages/groups.jsx
Comment (genehack) [reply static-site/src/pages/groups.jsx:16]: I refactored things so that the `<ListResources>` component gets passed a URL instead of a `sourceId`. I left the `fetch()` functionality where it is. With the way `useDataFetch` is currently implemented and structured, pulling the `fetch` call out into the resource listing call back would be fairly ugly, IMO.

The way things are structured now, `<ListResources>` fetches a URL and hands the response for that fetch off to the provided callback, which is responsible for producing an object with this shape: 

```json
{ 
  "pathPrefix": string,
  "pathVersions": (string|undefined)[],
}
```

I'm just not seeing where the `fetch` in the callback is an improvement over the updates I just pushed.

File: static-site/src/pages/groups.jsx
Comment (genehack) [reply static-site/src/pages/groups.jsx:24]: > For what it's worth, I'd use an empty array to represent no dates.

To be explicit: will there _ever_ be a resource that has *multiple* undated datasets? 

If the answer to that is "no", then sure, an empty array will work to indicate "no date".

If that answer to that is "...maybe", then I think an array with one `undefined` value per dataset is a more future proof approach.

File: static-site/src/components/ListResources/IndividualResource.tsx
Comment (genehack) [reply static-site/src/components/ListResources/IndividualResource.tsx]: Nope, this works great, changed.

**Comment by genehack** [comment]
Note: `CI / build` will be broken until [TS lint changes PR](https://github.com/nextstrain/nextstrain.org/pull/906) lands.

**Comment by victorlin** [comment]
> Note: `CI / build` will be broken until [TS lint changes PR](https://github.com/nextstrain/nextstrain.org/pull/906) lands.

This blockage doesn't seem necessary/productive. Attempting removal with #909

**Comment by genehack** [comment]
> > Note: `CI / build` will be broken until [TS lint changes PR](https://github.com/nextstrain/nextstrain.org/pull/906) lands.
> 
> This blockage doesn't seem necessary/productive. Attempting removal with #909

it's not _necessary_, in that i could add lines to disable the couple lint failures -- but then those would need to be removed once #906 lands, and that feels like unproductive churn to me. 

(i don't have a strong opinion about the direction #909 moves things in; i think linting being mildly annoying when broken is maybe a good thing…)

**Comment by victorlin** [comment]
@genehack continuing the tangential discussion in https://github.com/nextstrain/nextstrain.org/pull/909#issuecomment-2159141269

**Comment by genehack** [comment]
> @genehack continuing the tangential discussion in [#909 (comment)](https://github.com/nextstrain/nextstrain.org/pull/909#issuecomment-2159141269)

👍 

I merged #906 so that should get this unblocked soon (he says, as he discovers _another_ linting problem...) — I will link a preview link once it builds.

**Comment by genehack** [comment]
[Preview](https://nextstrain-s-group-reso-fafosr.herokuapp.com/groups) — note that only `blab` group data is being returned by the `charon/getAvailable` call in the staging/preview environment, and I'm not sure why that isn't reflected in the group tiles in the preview…

File: static-site/src/pages/groups.jsx
Comment (jameshadfield) [reply static-site/src/pages/groups.jsx:24]: If we had such data then the internal structure would have to change anyway because the UI would need a way to distinguish between them, ultimately by surfacing different URLs for each.

The difference between `[]` and `[undefined]` in JS is subtle and I don't know if I'd write code that relied on that subtlety.

Comment (genehack) [reply static-site/src/pages/groups.jsx:24]: > If we had such data then the internal structure would have to change anyway because the UI would need a way to distinguish between them, ultimately by surfacing different URLs for each.

okay, good to know. 

> The difference between `[]` and `[undefined]` in JS is subtle and I don't know if I'd write code that relied on that subtlety.

the main reason i would reach for `[undefined]` instead of `[]` is because all the existing code expects to iterate over a list, and `[undefined]` _can_ be iterated over. 

i will try to see how ugly an empty list will make this; i expect a couple parts will get more funky than they currently are — we'll see.

Comment (jameshadfield) [reply static-site/src/pages/groups.jsx:24]: `[undefined]` is fine by me

Comment (genehack) [reply static-site/src/pages/groups.jsx]: > i will try to see how ugly an empty list will make this; i expect a couple parts will get more funky than they currently are — we'll see.

was not horrible; pushed updates.

File: static-site/src/pages/groups.jsx
Comment (victorlin) [reply static-site/src/pages/groups.jsx:16]: > I'm just not seeing where the `fetch` in the callback is an improvement over the updates I just pushed.

The new updates are better, but I still see some benefits to having `fetch` in callback. Something like this, feel free to take or leave: c842d4d84c96af8eba3cb4c109f36e40c99ccca0

File: static-site/src/pages/groups.jsx
Comment (genehack) [reply static-site/src/pages/groups.jsx:111]: yep, good catch; should be fixed.

File: static-site/src/pages/groups.jsx
Comment (genehack) [reply static-site/src/pages/groups.jsx:20]: fixed.

File: static-site/src/components/ListResources/ResourceGroup.tsx
Comment (genehack) [reply static-site/src/components/ListResources/ResourceGroup.tsx:53]: 👍

File: static-site/src/components/ListResources/index.tsx
Comment (genehack) [reply static-site/src/components/ListResources/index.tsx:85]: 👍

File: static-site/src/components/ListResources/index.tsx
Comment (genehack) [reply static-site/src/components/ListResources/index.tsx]: fair! will breakout.

**Comment by genehack** [comment]
Replicating a f2f chat I had (separately) with both Victor and James:

> The new updates are better, but I still see some benefits to having fetch in callback. 

Originally I was thinking all the callbacks would be using `fetch()`, not understanding some would be static data structures. Knowing this, I agree there's some benefit to internalizing more responsibility into the callback. 

Reviewing Victor's draft version of this, the callbacks relying on importing a fetch helper/wrapper out of useDataFetch seemed a bit too pretzel-y to me. I'm going to try a version where the pathogens and staging pages duplicate the ~4-6 lines of fetch() logic, and where the general signature of the callback is it gets passed the `setDataError` state callback. I think that will ultimately loosen the coupling between the caller and the callback in a way that allows sufficient flexibility to cover all current, and many future, use cases.

File: static-site/src/pages/groups.jsx
Comment (genehack) [reply static-site/src/pages/groups.jsx]: folded those changes into the "replace datasets..." commit, thanks.

**Comment by genehack** [comment]
Okay, updated: callbacks now handle the fetch(); general other cleanups and fixups.

File: static-site/src/pages/groups.jsx
Comment (jameshadfield) [inline static-site/src/pages/groups.jsx:122]: [haven't reviewed, just noting something while playing with the review app]

Looks like the resource collapsing functionality isn't working (as in, the carot rotates and text changes, but the number of resources shown is unchanged). Or is this some conditional I added if number of cards is 1 or similar?

Comment (victorlin) [reply static-site/src/pages/groups.jsx:122]: It works for me (Vivaldi using Chrome/124.0.0.0):

https://github.com/nextstrain/nextstrain.org/assets/13424970/832c8db2-5eda-478a-994f-25f130e32410

Comment (jameshadfield) [reply static-site/src/pages/groups.jsx:122]: Oh. You've got a bigger screen than me ;) Maybe I set too big of a "show this many when collapsed" count - I remember I made it a function of the number of cards in play. Not a bug in this PR then.

Comment (victorlin) [reply static-site/src/pages/groups.jsx:122]: > You've got a bigger screen than me

Nah, I just zoomed out a bunch haha

File: static-site/src/components/ListResources/useDataFetch.ts
Comment (jameshadfield) [inline static-site/src/components/ListResources/useDataFetch.ts:32]: At the moment, if the callback _fails_ it's setting an error state and returning `undefined`, which is then causing a runtime error (because `const { pathPrefix, pathVersions } = undefined`) which gets caught and then... sets  the error state again¹ to `true`. 

Would it be simpler to just allow the callback to throw, and catch that here (which we are doing anyway)? If we want to set custom error messages then we need more sweeping changes which seem unnecessary at this stage.

¹ Conceptually speaking. In reality, I think the two `setDataFetchError` calls will be batched so you only see the last one. Maybe.

File: static-site/src/pages/groups.jsx
Comment (jameshadfield) [inline static-site/src/pages/groups.jsx:18]: FYI - the charon API doesn't use content negotiation. No need to change, but just so you know this header isn't doing anything ;)

File: static-site/src/components/ListResources/index.tsx
Comment (jameshadfield) [inline static-site/src/components/ListResources/index.tsx:83]: [not a request for changes - I just wanted to see if this was possible and am noting it here]

So you can use the elvis operator for array indexing too! `groups?.[0]?.lastUpdated && ...`

**Review by jameshadfield — COMMENTED** [review]
The overloading of the word "group" must have made this a confusing first experience!

The error handling seems a little convoluted to me and it seems like it could be simplified in this PR, but that's the only thing that stands out to me. Functionally it's working really well.

File: static-site/src/components/ListResources/useDataFetch.ts
Comment (genehack) [reply static-site/src/components/ListResources/useDataFetch.ts:32]: yeah, that's fair; i've ripped out the try/catch bits in the callbacks and it's clearer. will push update soon.

File: static-site/src/pages/groups.jsx
Comment (genehack) [reply static-site/src/pages/groups.jsx:18]: fixed since i was in there anyway.

File: static-site/src/components/ListResources/index.tsx
Comment (victorlin) [inline static-site/src/components/ListResources/index.tsx:164]: The change from `target` to `currentTarget` doesn't have any functional impact?

File: static-site/src/components/ListResources/useDataFetch.ts
Comment (victorlin) [inline static-site/src/components/ListResources/useDataFetch.ts:25]: > I'm going to try a version … where the general signature of the callback is it gets passed the `setDataError` state callback

I noticed you kept `setDataFetchError` in `useDataFetch` instead of the propsed implementation – not sure if you're planning to do that still but I think it looks good as-is!

**Review by victorlin — APPROVED** [review]

File: static-site/src/components/ListResources/useDataFetch.ts
Comment (victorlin) [reply static-site/src/components/ListResources/useDataFetch.ts:25]: Oh I didn't realize this was already discussed in https://github.com/nextstrain/nextstrain.org/pull/904#discussion_r1637489359, sorry

File: static-site/src/components/ListResources/index.tsx
Comment (genehack) [reply static-site/src/components/ListResources/index.tsx:164]: Not that I've noticed. The Typescript compiler tells me `event.target` is not defined, too.

**Comment by tsibley** [comment]
> note that only `blab` group data is being returned by the `charon/getAvailable` call in the staging/preview environment, and I'm not sure why that isn't reflected in the group tiles in the preview…

Maybe this was addressed in a meeting off-thread you had, but since I didn't see mention of it I wanted to clarify this for you.

Heroku review apps run in production mode with production config, _but_ we've intentionally configured them (on the Heroku side) to use the `nextstrain.org-testing` AWS IAM user, which has [a much more restrictive policy](https://github.com/nextstrain/nextstrain.org/blob/e70f5304ce0969d9e0214a043d9c3438df91ee2c/aws/iam/policy/NextstrainDotOrgServerInstance-testing.tftpl.json#L1) attached to it than the actual production IAM user (`nextstrain.org`)'s [attached policy](https://github.com/nextstrain/nextstrain.org/blob/e70f5304ce0969d9e0214a043d9c3438df91ee2c/aws/iam/policy/NextstrainDotOrgServerInstance.tftpl.json#L1).

**Comment by genehack** [comment]
> > note that only `blab` group data is being returned by the `charon/getAvailable` call in the staging/preview environment, and I'm not sure why that isn't reflected in the group tiles in the preview…
> 
> Maybe this was addressed in a meeting off-thread you had, but since I didn't see mention of it I wanted to clarify this for you.
> 
> Heroku review apps run in production mode with production config, _but_ we've intentionally configured them (on the Heroku side) to use the `nextstrain.org-testing` AWS IAM user, which has [a much more restrictive policy](https://github.com/nextstrain/nextstrain.org/blob/e70f5304ce0969d9e0214a043d9c3438df91ee2c/aws/iam/policy/NextstrainDotOrgServerInstance-testing.tftpl.json#L1) attached to it than the actual production IAM user (`nextstrain.org`)'s [attached policy](https://github.com/nextstrain/nextstrain.org/blob/e70f5304ce0969d9e0214a043d9c3438df91ee2c/aws/iam/policy/NextstrainDotOrgServerInstance.tftpl.json#L1).

nope, nobody else has shared this; thanks, this makes sense.

**Comment by trvrb** [comment]
Makes sense that this is just showing blab on the preview app, but the wider blab dataset list is a bit funny in terms of collapsing. The amount of content shown should be relative to the how many lines there are. Currently the blab dataset list is definitely too long even in collapsed view.

**Comment by genehack** [comment]
> Makes sense that this is just showing blab on the preview app, but the wider blab dataset list is a bit funny in terms of collapsing. The amount of content shown should be relative to the how many lines there are. Currently the blab dataset list is definitely too long even in collapsed view.

I will look at what's going on here. Interestingly, if I point my local dev instance at the production data endpoint (i.e., if I fetch the data fram `https://nextstrain.org/charon/getAvailable?prefix=/groups/`), this does not happen: 

Screenshot 2024-06-14 at 10 10 06

I wonder if the oddness in the display is because of the smaller data set size available in the preview environment?

**Comment by genehack** [comment]
> I wonder if the oddness in the display is because of the smaller data set size available in the preview environment?

Experimenting further, if I take the production data but filter it down to just 2 cards worth of data, the cards I get are large. Expanding that to 3 cards, the layout is as expected. 

Two groups:

Screenshot 2024-06-14 at 10 23 55

Three groups: 

Screenshot 2024-06-14 at 10 24 09

**Comment by genehack** [comment]
> > I wonder if the oddness in the display is because of the smaller data set size available in the preview environment?
> 
> Experimenting further, if I take the production data but filter it down to just 2 cards worth of data, the cards I get are large. Expanding that to 3 cards, the layout is as expected.

Digging ever further, [this seems to be intentional behavior](https://github.com/nextstrain/nextstrain.org/blob/master/static-site/src/components/ListResources/ResourceGroup.tsx#L267-L279). @jameshadfield could you speak to this? 

I'm going to say I don't think this should be something that blocks this PR; if we do want to change this behavior, it can be a distinct thing.

**Comment by jameshadfield** [comment]
>Digging ever further, [this seems to be intentional behavior](https://github.com/nextstrain/nextstrain.org/blob/master/static-site/src/components/ListResources/ResourceGroup.tsx#L267-L279). @jameshadfield could you speak to this?

The intention is that if few cards are in view (commonly achieved via filtering) then we should show more entries per card. The exact thresholds are (as always) open to change. We could (probably should) also remove the expand/contract toggle if there is only 1 card.

**Pull request merged**
